File: c++/maybe.cc
#include <stdexcept>
#include <random>
#include "maybe.h"

unsigned int Maybe::maybes = 0;

Maybe::Maybe()
  :truth{0.5} {
  maybetrue = this->maybe();
}

Maybe::Maybe(float truthp){
  if (truthp <= 0.0 || truthp >= 1.0) {
    throw std::invalid_argument("A maybe object must have a truth percentage between 0.0 and 1.0");
  }
  truth = truthp;
  maybetrue = this->maybe();
}

bool Maybe::maybe(){
  maybes++;
  bool tof{false};
  float n{randomizer()};
  
  if (n < truth) {
    tof = true;
  }
  
  return tof;
}

bool Maybe::is_true(){
  return maybetrue;
}

bool Maybe::is_false(){
  return ( maybetrue == false) ;
}

int Maybe::uncertain(){
  return maybes;
}

void Maybe::update() {
  maybetrue = this->maybe();
}

void Maybe::update(float truthp) {
  if (truthp <= 0.0 || truthp >= 1.0) {
    throw std::invalid_argument("A maybe object must have a truth percentage between(or equal to) 0.0 and 1.0");
  }
  truth = truthp;
  maybetrue = this->maybe();
}

float Maybe::randomizer() {
  std::default_random_engine generator;
  std::uniform_real_distribution<float> distribution(0.0,1.0);
  float n = distribution(generator);	
  while (n == truth)
  {
	n = distribution(generator);
  }
	
    return n;
}

File: Python/maybe.py
from random import randint

class Maybe(object):

    maybes = 0

    def maybe_(self):
        Maybe.maybes += 1

        n = randint(0,100)
        while (n == 50):
            n = randint(0,100)
        n = n/100.0
        if n < self.truthPercent:
            return True
        else:
            return False

    def update(self, truthP=None):
        if (truthP == None):
            pass
        elif ( truthP >= 0.0  and truthP <= 1.0 ):
            self.truthPercent = truthP
        else: 
            raise AttributeError("A maybe's truthPercent must be between 0.0 and 1.0")
            
        self.maybe = self.maybe_()
    

    def __init__(self, truthP=0.5):
        if ( (truthP >= 0.0) and (truthP <= 1.0) ):
            self.truthPercent = truthP 
        else: 
            raise AttributeError("A maybe's truthPercent must be between 0.0 and 1.0")
        
        self.maybe = self.maybe_()

    def __bool__(self):
        return self.maybe

    def __eq__(self, other):
        if isinstance(other,Maybe):
            return self.maybe == other.maybe
        else:
            return self.maybe == other
        
    def __ne__(self, other):
        if isinstance(other,Maybe):
            return self.maybe != other.maybe
        else:
            return self.maybe != other
